File: analytics/src/test/java/io/appmetrica/analytics/impl/startup/uuid/UuidFromStartupStateImporterTest.java
package io.appmetrica.analytics.impl.startup.uuid;

import android.content.Context;
import io.appmetrica.analytics.coreapi.internal.data.ProtobufStateStorage;
import io.appmetrica.analytics.impl.db.state.factory.StorageFactory;
import io.appmetrica.analytics.impl.startup.StartupState;
import io.appmetrica.analytics.testutils.CommonTest;
import io.appmetrica.analytics.testutils.MockedStaticRule;
import java.util.UUID;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.robolectric.RobolectricTestRunner;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.when;

@RunWith(RobolectricTestRunner.class)
public class UuidFromStartupStateImporterTest extends CommonTest {

    @Mock
    private Context context;
    @Mock
    private StorageFactory<StartupState> startupStateStorageFactory;
    @Mock
    private ProtobufStateStorage<StartupState> protobufStateStorage;
    @Mock
    private StartupState startupState;

    @Rule
    public MockedStaticRule<StorageFactory.Provider> storageFactoryProviderMockedStaticRule =
        new MockedStaticRule<>(StorageFactory.Provider.class);

    private UuidFromStartupStateImporter uuidFromStartupStateImporter;

    @Before
    public void setUp() throws Exception {
        MockitoAnnotations.openMocks(this);

        when(StorageFactory.Provider.get(StartupState.class)).thenReturn(startupStateStorageFactory);
        when(startupStateStorageFactory.create(context)).thenReturn(protobufStateStorage);
        when(protobufStateStorage.read()).thenReturn(startupState);

        uuidFromStartupStateImporter = new UuidFromStartupStateImporter();
    }

    @Test
    public void get() {
        String uuid = UUID.randomUUID().toString();

        when(startupState.getUuid()).thenReturn(uuid);

        assertThat(uuidFromStartupStateImporter.get(context)).isEqualTo(uuid);
    }

    @Test
    public void getIfUuidDoesNotExist() {
        when(startupState.getUuid()).thenReturn(null);

        assertThat(uuidFromStartupStateImporter.get(context)).isNull();
    }
}

File: analytics/src/test/java/io/appmetrica/analytics/impl/startup/uuid/UuidFromStartupStateImporterTest.kt
package io.appmetrica.analytics.impl.startup.uuid

import android.content.Context
import io.appmetrica.analytics.coreapi.internal.data.ProtobufStateStorage
import io.appmetrica.analytics.impl.db.state.factory.StorageFactory
import io.appmetrica.analytics.impl.startup.CollectingFlags.CollectingFlagsBuilder
import io.appmetrica.analytics.impl.startup.StartupStateModel
import io.appmetrica.analytics.impl.startup.StartupStateModel.StartupStateBuilder
import io.appmetrica.analytics.testutils.CommonTest
import io.appmetrica.analytics.testutils.on
import io.appmetrica.analytics.testutils.staticRule
import org.assertj.core.api.Assertions.assertThat
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.kotlin.doReturn
import org.mockito.kotlin.mock
import org.mockito.kotlin.whenever
import org.robolectric.RobolectricTestRunner
import java.util.UUID

@RunWith(RobolectricTestRunner::class)

class UuidFromStartupStateImporterTest : CommonTest() {

    private val context: Context = mock()

    private val protobufStateStorage: ProtobufStateStorage<StartupStateModel> = mock()

    private val startupStateStorageFactory: StorageFactory<StartupStateModel> = mock {
        on { create(context) } doReturn protobufStateStorage
    }

    @get:Rule
    val storageFactoryProviderMockedStaticRule = staticRule<StorageFactory.Provider> {
        on { StorageFactory.Provider.get(StartupStateModel::class.java) } doReturn startupStateStorageFactory
    }
    private val uuidFromStartupStateImporter: UuidFromStartupStateImporter by setUp { UuidFromStartupStateImporter() }

    @Test
    fun `get if uuid is exist`() {
        val uuid = UUID.randomUUID().toString()
        whenever(protobufStateStorage.read()).thenReturn(
            StartupStateBuilder(CollectingFlagsBuilder().build())
                .withUuid(uuid)
                .build()
        )
        assertThat(uuidFromStartupStateImporter[context]).isEqualTo(uuid)
    }

    @Test
    fun `get if uuid is not exist`() {
        whenever(protobufStateStorage.read()).thenReturn(StartupStateBuilder(CollectingFlagsBuilder().build()).build())
        assertThat(uuidFromStartupStateImporter[context]).isNull()
    }

    @Test
    fun `get if factory is missing`() {
        whenever(StorageFactory.Provider.get(StartupStateModel::class.java)).thenReturn(null)
        assertThat(uuidFromStartupStateImporter[context]).isNull()
    }

    @Test
    fun `get if throw`() {
        whenever(protobufStateStorage.read()).thenThrow(RuntimeException())
        assertThat(uuidFromStartupStateImporter[context]).isNull()
    }
}
